File: c++/source/morth/op.cpp
#include "morth/op.hpp"

Op Push(Int64 value, Token tok) {
    Op op;
    op.code = OpCode::PUSH;
    op.value = value;
    op.tok = tok;
    return op;
}

Op Plus(Token tok) {
    Op op;
    op.code = OpCode::PLUS;
    op.tok = tok;
    return op;
}

Op Minus(Token tok) {
    Op op;
    op.code = OpCode::MINUS;
    op.tok = tok;
    return op;
}

Op Dump(Token tok) {
    Op op;
    op.code = OpCode::DUMP;
    op.tok = tok;
    return op;
}

File: fs/Morth.Language/Op.fs
module Morth.Language.Op

type code =
  | Push of int
  | Dup
  | Plus
  | Minus
  | Eq
  | Gt
  | Dump
  | If of int
  | Else of int
  | While
  | Do of int
  | End of int
  | Mem
  | Load
  | Store
  | Syscall1
  | Syscall2
  | Syscall3
  | Syscall4
  | Syscall5
  | Syscall6

type t =
  struct
    val loc : Token.loc
    val mutable code : code

    new(loc : Token.loc, code : code) = { loc = loc; code = code }
  end
